xbmc/utils/Observer.cpp

```diff
@@ -98,44 +46,40 @@ void Observable::RegisterObserver(Observer *obs)
   if (!IsObserving(*obs))
   {
     m_observers.push_back(obs);
-    obs->RegisterObservable(this);
   }
 }
 
 void Observable::UnregisterObserver(Observer *obs)
 {
   CSingleLock lock(m_obsCritSection);
-  std::vector<Observer *>::iterator it = find(m_observers.begin(), m_observers.end(), obs);
-  if (it != m_observers.end())
-  {
-    obs->UnregisterObservable(this);
-    m_observers.erase(it);
-  }
+  std::remove(m_observers.begin(), m_observers.end(), obs);
 }
 
 void Observable::NotifyObservers(const ObservableMessage message /* = ObservableMessageNone */)
 {
-  bool bNotify(false);
-  {
-    CSingleLock lock(m_obsCritSection);
-    if (m_bObservableChanged)
-      bNotify = true;
-    m_bObservableChanged = false;
-  }
+  // Make sure the set/compare is atomic 
+  // so we don't clobber the variable in a race condition
+  auto bNotify = m_bObservableChanged.exchange(false);
 
   if (bNotify)
     SendMessage(message);
 }
 
 void Observable::SetChanged(bool SetTo)
 {
-  CSingleLock lock(m_obsCritSection);
   m_bObservableChanged = SetTo;
 }
 
 void Observable::SendMessage(const ObservableMessage message)
 {
   CSingleLock lock(m_obsCritSection);
+  auto obs = m_observers;
```

Comment:
you introduce an unused variable here.
I understand what you have in mind but this would only work with shared pointers. If an observer unregisters after leaving the lock, this segfaults.